Overhaul Benchmarking pipeline to use complete sample data, not summaries

tbkka opened the pull request:
With this change, the Swift benchmarking harness now has two distinct output formats:

* Default: Formatted text that's intended for human consumption. Right now, this is just the minimum value, but we can augment that.

* `--json`: each output line is a JSON-encoded object that contains raw data This information is intended for use by python scripts that aggregate or compare multiple independent tests.

Previously, we tried to use the same output for both purposes.  This required the python scripts to do more complex parsing of textual layouts, and also meant that the python scripts had only summary data to work with instead of full raw sample information.  This in turn made it almost impossible to derive meaningful comparisons between runs or to aggregate multiple runs.

Typical output in the new JSON format looks like this:
```
{"number":89, "name":"PerfTest", "samples":[1.23, 2.35], "max_rss":16384}
{"number":91, "name":"OtherTest", "samples":[14.8, 19.7]}
```
Note the `"samples"` array includes the runtime for each individual run.

This format is easy to parse in Python.  Just iterate over lines and decode each one separately. Also note that the optional fields (`"max_rss"` above) are trivial to handle:
```
import json
for l in lines:
   j = json.loads(l)
   # Default 0 if not present
   max_rss = j.get("max_rss", 0)
```

Because optional fields are so much easier to handle in this form, I reworked the Python logic to translate old formats into this JSON format for more uniformity.  Hopefully, we can simplify the code in a year or so by stripping out the old log formats entirely, along with some of the redundant statistical calculations.  In particular, the python logic still makes an effort to preserve mean, median, max, min, stdev, and other statistical data whenever the full set of samples is not present.  Once we've gotten to a point where we're always keeping full samples, we can compute any such information on the fly as needed, eliminating the need to record it.

This is a pretty big rearchitecture of the core benchmarking logic.  In order to try to keep things a bit more manageable, I have _not_ taken this opportunity to replace any of the actual statistics used in the higher level code or to change how the actual samples are measured. (But I expect this rearchitecture will make such changes simpler.) In particular, this should not actually change any benchmark results.

For the future, please keep this general principle in mind: Statistical summaries (averages, medians, etc) should as a rule be computed for immediate output and rarely if ever stored or used as input for other processing. Instead, aim to store and transfer raw data from which statistics can be recomputed as necessary.

tbkka committed:
Overhaul Benchmarking pipeline to use complete sample data, not summaries

The Swift benchmarking harness now has two distinct output formats:

* Default: Formatted text that's intended for human consumption.
  Right now, this is just the minimum value, but we can augment that.

* `--json`: each output line is a JSON-encoded object that contains raw data
  This information is intended for use by python scripts that aggregate
  or compare multiple independent tests.

Previously, we tried to use the same output for both purposes.  This required
the python scripts to do more complex parsing of textual layouts, and also meant
that the python scripts had only summary data to work with instead of full raw
sample information.  This in turn made it almost impossible to derive meaningful
comparisons between runs or to aggregate multiple runs.

Typical output in the new JSON format looks like this:
```
{"number":89, "name":"PerfTest", "samples":[1.23, 2.35], "max_rss":16384}
{"number":91, "name":"OtherTest", "samples":[14.8, 19.7]}
```

This format is easy to parse in Python.  Just iterate over
lines and decode each one separately. Also note that the
optional fields (`"max_rss"` above) are trivial to handle:
```
import json
for l in lines:
   j = json.loads(l)
   # Default 0 if not present
   max_rss = j.get("max_rss", 0)
```
Note the `"samples"` array includes the runtime for each individual run.

Because optional fields are so much easier to handle in this form, I reworked
the Python logic to translate old formats into this JSON format for more
uniformity.  Hopefully, we can simplify the code in a year or so by stripping
out the old log formats entirely, along with some of the redundant statistical
calculations.  In particular, the python logic still makes an effort to preserve
mean, median, max, min, stdev, and other statistical data whenever the full set
of samples is not present.  Once we've gotten to a point where we're always
keeping full samples, we can compute any such information on the fly as needed,
eliminating the need to record it.

This is a pretty big rearchitecture of the core benchmarking logic. In order to
try to keep things a bit more manageable, I have not taken this opportunity to
replace any of the actual statistics used in the higher level code or to change
how the actual samples are measured. (But I expect this rearchitecture will make
such changes simpler.) In particular, this should not actually change any
benchmark results.

For the future, please keep this general principle in mind: Statistical
summaries (averages, medians, etc) should as a rule be computed for immediate
output and rarely if ever stored or used as input for other processing. Instead,
aim to store and transfer raw data from which statistics can be recomputed as
necessary.

tbkka commented:
@swift-ci Please test

atrick reviewed (commented):
Much appreciated. I can't think of any issues with this change.

shahmishal commented:
The new json output looks great, thanks!

tbkka committed:
Fix some test failures

The new code stores test numbers as numbers (not strings), which
requires a few adjustments. I also apparently missed a few test updates.

shahmishal commented:
@swift-ci Python lint

shahmishal commented:
@swift-ci smoke benchmark

tbkka committed:
Python style fixes

tbkka committed:
Comment some TODO items

tbkka committed:
Pylint cleanup, more comments

tbkka commented:
@swift-ci Please benchmark

tbkka commented:
@swift-ci benchmark

tbkka committed:
Do not use --json for listing tests (yet)

tbkka committed:
Fix underflow in the padding calculation

tbkka commented:
@swift-ci Please benchmark

tbkka commented:
@swift-ci benchmark

tbkka committed:
Fix colliding fields; match old format more closely

tbkka committed:
Match new benchmark driver default output

tbkka committed:
Use non-json format for now until we have switched over completely

tbkka commented:
@swift-ci benchmark

tbkka committed:
Make the default output a little more like the old version (for now)

tbkka commented:
@swift-ci benchmark

tbkka committed:
Use results consistently

tbkka committed:
For size comparisons, build the result objects directly with sample data

tbkka committed:
Make --num-samples actually work

tbkka commented:
@swift-ci benchmark

tbkka commented:
@swift-ci test

tbkka committed:
A better way to adapt to -num-samples

tbkka commented:
@swift-ci Please test

tbkka committed:
Pylint fixes

tbkka committed:
pylint fixes

tbkka committed:
Test the non-JSON output

We have to continue using the non-JSON forms
until the JSON-supporting code is universally
available.

tbkka commented:
Some of the users of the benchmark harness rely on being able to use previously-compiled benchmark binaries.  So I've switched things back over so the Python scripts request the non-JSON format.  In a month or so when the JSON-supporting binaries are generally available everywhere, I'll switch it back.  This way, benchmarking should continue to work throughout the transition period.

tbkka commented:
@swift-ci Please test

tbkka commented:
@swift-ci benchmark

tbkka commented:
@swift-ci Please benchmark

tbkka commented:
Another improvement I slipped in:  Sample data is now computed in floating point rather than integer microseconds, so we can now correctly handle scaled results less than 1µs.  This shows up as some amusing "regressions" compared to the old benchmark harness:
```
SubstringRemoveLast1                  0.0     0.196     +19600.0%   **0.01x**
SubstringRemoveFirst1                 0.0     0.167     +16700.0%   **0.01x (?)**
StringWithCString2                    0.0     0.006     +600.0%     **0.14x**
```
